woocommerce/woocommerce-ios
# Set order's customer ID from selecting a customer during order creation

Closes: #8257 


## Why


In order creation, the merchant can add an existing customer to prefill the addresses. If the customer is a registered user of the store, the customer should be linked from the order but it currently isn't. This PR links the customer to the order by setting `customer_id` in the order. This change can help with customer management for analytics and marketing.

## How

In `OrdersRemote`, `customerID` was added to the order creation/update field enums so that the customer ID can be set in the order request body.

In `RemoteOrderSynchronizer` which coordinates how an order field update is handled (whether to sync remotely in order creation/editing), `setCustomerID` was added so that it can update the latest order's customer ID for remote sync later. Because the customer's ID and addresses are set at the same time, the customer ID is set first locally and then setting addresses will trigger a remote sync.

Finally, to integrate with the customer selection flow in order creation, `setCustomerID` is called in `EditableOrderViewModel.addCustomerAddressToOrder` when the merchant selects a customer from the selector. Please note that only registered customers can be associated with an order.

## Testing instructions


Prerequisite: the store enables account creation, the settings are in `/admin.php?page=wc-settings&tab=account` (somehow account registration is disabled by default in my Woo Express test store). Then, create a customer account if there are no registered customers yet.

- Go to the orders tab
- Tap + to create an order
- Tap `Add customer details`
- Select a registered customer (the user ID is shown to the trailing edge of the customer name)
- Add any other details to the order if you like
- Tap `Create` --> the order should be created remotely, and the customer should be set in wp-admin

## Screenshots



https://github.com/woocommerce/woocommerce-ios/assets/1945542/7a4f1516-96ec-498a-9d38-b944f940a4cf

<img width="447" alt="Screenshot 2024-01-09 at 12 55 57 PM" src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/3780148b-5745-46fc-950b-b2d81c1c652c">


---
